File: src/algebra1_factoring_gcf_medium.py
"""
algebra1 - factoring_gcf (medium)
Generated: 2026-02-22T04:00:06.234293
"""

from problem_utils import *

def generate():
    # Generate GCF factoring problems at 1300-1600 difficulty
    # Standard level: 2-3 terms with clear GCF
    
    problem_type = randint(1, 3)
    
    if problem_type == 1:
        # Two-term factoring with monomial GCF (1300-1400)
        gcf_coeff = randint(2, 6)
        gcf_power = randint(1, 3)
        
        term1_extra_coeff = randint(2, 8)
        term1_extra_power = randint(0, 2)
        
        term2_extra_coeff = randint(2, 8)
        term2_extra_power = randint(0, 2)
        
        # Build terms
        term1 = gcf_coeff * term1_extra_coeff * x**(gcf_power + term1_extra_power)
        term2 = gcf_coeff * term2_extra_coeff * x**(gcf_power + term2_extra_power)
        
        expr = term1 + term2
        ans = factor(expr)
        
        difficulty_val = (1300, 1400)
        sol = steps(
            f"Identify the GCF of ${latex(term1)}$ and ${latex(term2)}$",
            f"GCF = ${latex(gcf_coeff * x**gcf_power)}$",
            f"Factor out the GCF: ${latex(ans)}$"
        )
        
    elif problem_type == 2:
        # Three-term factoring with monomial GCF (1400-1500)
        gcf_coeff = randint(2, 5)
        gcf_power = randint(1, 2)
        
        term1_extra_coeff = randint(1, 6)
        term1_extra_power = randint(1, 3)
        
        term2_extra_coeff = randint(1, 6)
        term2_extra_power = randint(0, 2)
        
        term3_extra_coeff = randint(1, 6)
        
        # Build terms
        term1 = gcf_coeff * term1_extra_coeff * x**(gcf_power + term1_extra_power)
        term2 = gcf_coeff * term2_extra_coeff * x**(gcf_power + term2_extra_power)
        term3 = gcf_coeff * term3_extra_coeff * x**gcf_power
        
        expr = term1 + term2 + term3
        ans = factor(expr)
        
        difficulty_val = (1400, 1500)
        sol = steps(
            f"Identify the GCF of all terms in ${latex(expr)}$",
            f"GCF = ${latex(gcf_coeff * x**gcf_power)}$",
            f"Factor out the GCF: ${latex(ans)}$"
        )
        
    else:
        # Three-term with negative coefficient and higher powers (1500-1600)
        gcf_coeff = randint(2, 4)
        gcf_power = randint(1, 2)
        
        # Make first coefficient negative for added complexity
        sign = -1
        
        term1_extra_coeff = randint(2, 5)
        term1_extra_power = randint(2, 3)
        
        term2_extra_coeff = randint(2, 5)
        term2_extra_power = randint(1, 2)
        
        term3_extra_coeff = randint(1, 4)
        
        # Build terms
        term1 = sign * gcf_coeff * term1_extra_coeff * x**(gcf_power + term1_extra_power)
        term2 = sign * gcf_coeff * term2_extra_coeff * x**(gcf_power + term2_extra_power)
        term3 = sign * gcf_coeff * term3_extra_coeff * x**gcf_power
        
        expr = term1 + term2 + term3
        ans = factor(expr)
        
        difficulty_val = (1500, 1600)
        sol = steps(
            f"Identify the GCF of all terms in ${latex(expr)}$",
            f"Note: Factor out the negative to simplify",
            f"GCF = ${latex(sign * gcf_coeff * x**gcf_power)}$",
            f"Factor out the GCF: ${latex(ans)}$"
        )
    
    return problem(
        question=f"Factor out the greatest common factor (GCF): ${latex(expr)}$",
        answer=ans,
        difficulty=difficulty_val,
        topic="algebra1/factoring_gcf",
        grading_mode="factor",
        solution=sol
    )

emit(generate())

File: src/algebra1_factoring_gcf_medium.jl
# algebra1 - factoring_gcf (medium)
# Generated: 2026-03-08T19:59:49.232006

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x begin
    set_topic!("algebra1/factoring_gcf")
    
    # Problem type selection
    prob_type = rand(1:4)
    
    if prob_type == 1
        # Simple GCF factoring: gcf * (polynomial)
        gcf = nonzero(-12, 12)
        n_terms = rand(2:3)
        
        if n_terms == 2
            c1 = nonzero(-8, 8)
            c2 = nonzero(-8, 8)
            exp1 = rand(1:4)
            exp2 = rand(1:exp1-1)
            
            inside = c1 * x^exp1 + c2 * x^exp2
            expr = expand(gcf * inside)
            factored = gcf * inside
        else
            c1 = nonzero(-8, 8)
            c2 = nonzero(-8, 8)
            c3 = nonzero(-8, 8)
            exp1 = rand(2:4)
            exp2 = rand(1:exp1-1)
            
            inside = c1 * x^exp1 + c2 * x^exp2 + c3
            expr = expand(gcf * inside)
            factored = gcf * inside
        end
        
        diff = (1200, 1400)
        sol_steps = steps(
            sol("Given", expr),
            "Find GCF of all terms: $(abs(gcf))",
            "Factor out the GCF",
            sol("Answer", factored)
        )
        
    elseif prob_type == 2
        # GCF with higher powers and more terms
        gcf = nonzero(-15, 15)
        c1 = nonzero(-6, 6)
        c2 = nonzero(-6, 6)
        c3 = nonzero(-6, 6)
        c4 = nonzero(-6, 6)
        
        exp1 = rand(3:6)
        exp2 = rand(2:exp1-1)
        exp3 = rand(1:exp2-1)
        
        inside = c1 * x^exp1 + c2 * x^exp2 + c3 * x^exp3 + c4
        expr = expand(gcf * inside)
        factored = gcf * inside
        
        diff = (1400, 1600)
        sol_steps = steps(
            sol("Given", expr),
            "Identify GCF of coefficients: $(abs(gcf))",
            "Factor out the GCF from each term",
            sol("Factored form", factored)
        )
        
    elseif prob_type == 3
        # Negative leading coefficient that should be factored out
        gcf = -1 * rand(2:10)
        c1 = rand(2:8)  # Make inside have positive leading term
        c2 = nonzero(-10, 10)
        c3 = nonzero(-10, 10)
        
        exp1 = rand(2:4)
        exp2 = rand(1:exp1-1)
        
        inside = c1 * x^exp1 + c2 * x^exp2 + c3
        expr = expand(gcf * inside)
        factored = gcf * inside
        
        diff = (1300, 1500)
        sol_steps = steps(
            sol("Given", expr),
            "Factor out $(gcf) to make leading coefficient positive",
            "Distribute the negative GCF",
            sol("Answer", factored)
        )
        
    else
        # Large GCF with complex polynomial
        gcf_base = rand([2, 3, 5, 6, 10, 12, 15])
        gcf_mult = rand(2:4)
        gcf = gcf_base * gcf_mult
        
        c1 = nonzero(-5, 5)
        c2 = nonzero(-5, 5)
        c3 = nonzero(-5, 5)
        
        exp1 = rand(3:5)
        exp2 = rand(2:exp1-1)
        exp3 = rand(1:exp2-1)
        
        inside = c1 * x^exp1 + c2 * x^exp2 + c3 * x^exp3
        expr = expand(gcf * inside)
        factored = gcf * inside
        
        diff = (1500, 1700)
        sol_steps = steps(
            sol("Given", expr),
            "Find GCF of all coefficients: $(gcf)",
            "Factor out $(gcf) from each term",
            sol("Factored form", factored)
        )
    end
    
    problem(
        question="Factor completely by finding the GCF: \$$(tex(expr))\$",
        answer=factored,
        difficulty=diff,
        grading_mode="factor",
        solution=sol_steps,
        time=120
    )
end